### ISSUE TITLE
Nostd entrypoint (less CU)

### ISSUE THREAD

**Issue by ebrightfield** [regolith-labs/steel#1]
I haven't tested it much, but someone wrote this awesome [solana-nostd-entrypoint](https://github.com/cavemanloverboy/solana-nostd-entrypoint/tree/main) crate.

I just played around with a feature-flagged re-export of it in steel, and an example using it [here](https://github.com/ebrightfield/steel/commit/3fc6277920c1dfedcf3361c87b66bc67c0397a4a). 

Supposedly it uses much less CU than the "vanilla" Solana entrypoint macro (I haven't personally benchmarked it yet). The main difference that it introduces is replacing `AccountInfo<'_>` with a `NoStdAccountInfo`, which in turn required minor changes to the loaders (implemented in the commit linked above), and CPIs (not implemented in the commit yet). 

In the spirit of optimization, it could be a good choice for steel!

**Comment by HardhatChad** [comment]
Yes, I have discussed this with Cavey at length. I would like to introduce that here and would approve any PR that does so. As you mentioned, the main complications are just replacing `AccountInfo<'_>` and using `invoke_signed_c` for CPIs. When I was working on ORE, the nostd entrypoint crate was brand new (still being written) so I avoided using it at the time. I think to integrate it here, one should add a `"nostd"` feature flag and optionally reimplement that basic features using the types from the nostd crate.

**Comment by Jac0xb** [comment]
For what its worth replace AccountInfo is not easy to replace since there are many places in the solana program crate that use it. I tried to write a solana program with this entry point and once you start trying to do instruction building for CPI it becomes a chore.

**Comment by ebrightfield** [comment]
Yeah the CPI stuff definitely is the biggest challenge here, updating the loaders was a no-brainer, building instructions for CPI is definitely a chore, as illustrated by the [example](https://github.com/cavemanloverboy/solana-nostd-entrypoint/blob/d7cb3cd88464f1f470a00183291aa8472ea28651/example-program/src/lib.rs#L36) in the nostd-entrypoint repo.

Thinking that an IDL-assisted instruction building interface might be the missing piece here. The `.to_info_c()` and `.to_meta_c()` boilerplate can be routinized, as long as the instruction interface is well-defined up-front somewhere.
